cpp/src/arrow/device.cc

```diff
@@ -268,4 +270,67 @@ std::shared_ptr<MemoryManager> CPUDevice::default_memory_manager() {
   return default_cpu_memory_manager();
 }
 
+namespace internal {
+
+class DeviceMemoryManagerRegistryImpl {
+ public:
+  DeviceMemoryManagerRegistryImpl() {}
+
+  Status RegisterDevice(DeviceAllocationType device_type, MemoryMapper memory_mapper) {
+    std::lock_guard<std::mutex> lock(lock_);
+    auto it = registry_.find(device_type);
+    if (it != registry_.end()) {
+      return Status::KeyError("Device type ", static_cast<int>(device_type),
+                              " is already registered");
+    }
+    registry_[device_type] = std::move(memory_mapper);
```

Comment:
This is fine, but it can perhaps be shortened using https://en.cppreference.com/w/cpp/container/unordered_map/try_emplace :
```suggestion
    auto [_, inserted] = registry.try_emplace(device_type, std::move(memory_mapper));
    if (!inserted) {
      return Status::KeyError("Device type ", static_cast<int>(device_type),
                              " is already registered");
    }
```